controller/backing_image_controller.go

```diff
@@ -310,22 +345,604 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
+		}
+
+		delete(bi.Status.DiskFileStatusMap, v2DiskUUID)
+		bic.v2CopyBackoff.DeleteEntry(v2DiskUUID)
+		bic.eventRecorder.Eventf(bi, corev1.EventTypeNormal, constant.EventReasonDelete, "Deleted failed v2 backing image %v on disk %v on node %v", bi.Name, v2DiskUUID, node.Name)
+	}
+
+	// When the backing image is created on that disk, it is either ready, failed or unknown.
+	// We directly delete the failed v2 copy if there is ready copy
+	// For unknown v2 copy, it can happen in the following case,
+	// 1. If the instance-manager pod is not running or the instance-manager CR is not ready.
+	// 2. Backing image copy has status (ready or failed) before but can't get the ststus now.
+	// 3. Backing image copy is failed to be created on the disk and there is no lvol snapshot and at the same time the node is rebooted.
+	//    It will keeps in unknown state forever because spdk server has no snapshot lvol to pick up to reconstruct the record.
+	// For case 1 and case 2, the status will come back automatically when server pick the snapshot lvol up and reconstruct the record.
+	// We delete the unknown copy only for the case 3 if there is a ready copy so it can be recreated.
+	if !hasReadyV2BackingImage {
+		return nil
+	}
+
+	for _, v2DiskUUID := range v2CopiesUnknown {
+		node, _, err := bic.ds.GetReadyDiskNodeRO(v2DiskUUID)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the ready disk node for disk %v", v2DiskUUID)
+		}
+
+		instanceManager, err := bic.ds.GetRunningInstanceManagerByNodeRO(node.Name, longhorn.DataEngineTypeV2)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the default instance manager for node %v", node.Name)
+		}
+		engineClientProxy, err := engineapi.NewEngineClientProxy(instanceManager, log, bic.proxyConnCounter)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the engine client proxy for instance manager %v", instanceManager.Name)
+		}
+
+		// Get will return different error if lvol is found on the disk but the record is not found.
+		// We don't delete the v2 copy in this case
+		_, err = engineClientProxy.SPDKBackingImageGet(bi.Name, v2DiskUUID)
+		if err != nil {
+			if types.ErrorRecordNotFoundButLvolFound(err) {
+				log.Infof("Unknown v2 Copy on disk %v has lvol but no record yet, wait for the server to reconstruct the record.", v2DiskUUID)
+				continue
+			}
+		}
+
+		log.Infof("Deleting the unknown v2 copy on disk %v", v2DiskUUID)
+		err = engineClientProxy.SPDKBackingImageDelete(bi.Name, v2DiskUUID)
+		engineClientProxy.Close()
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete the v2 copy on diskUUID %v", v2DiskUUID)
+		}
+		delete(bi.Status.DiskFileStatusMap, v2DiskUUID)
+		bic.v2CopyBackoff.DeleteEntry(v2DiskUUID)
+		bic.eventRecorder.Eventf(bi, corev1.EventTypeNormal, constant.EventReasonDelete, "Deleted failed v2 backing image %v on disk %v on node %v", bi.Name, v2DiskUUID, node.Name)
```

Comment:
Suggest abstracting the codes into a function and close the client in a defer function.

Reply:
fixed.